src/plot_api/plot.js

```diff
@@ -0,0 +1,1483 @@
+                    .remove();
+            }
+        }
+
+        for (i = 0; i < subplots.length; i++) {
+            subplot = subplots[i];
+            subplotInfo = gd._fullLayout._plots[subplot];
+            cdSubplot = getCdSubplot(calcdata, subplot);
+            cdError = [];
+
+            // remove old traces, then redraw everything
+            // TODO: use enter/exit appropriately in the plot functions
+            // so we don't need this - should sometimes be a big speedup
+            if(subplotInfo.plot) subplotInfo.plot.selectAll('g.trace').remove();
+
+            for(j = 0; j < modules.length; j++) {
+                module = modules[j];
+                if(!module.plot) continue;
+
+                // plot all traces of this type on this subplot at once
+                cdModule = getCdModule(cdSubplot, module);
+                module.plot(gd, subplotInfo, cdModule);
+                Lib.markTime('done ' + (cdModule[0] && cdModule[0][0].trace.type));
+
+                // collect the traces that may have error bars
+                if(cdModule[0] && cdModule[0][0].trace && Plots.traceIs(cdModule[0][0].trace, 'errorBarsOK')) {
+                    cdError = cdError.concat(cdModule);
+                }
+            }
+
+            // finally do all error bars at once
+            if(gd._fullLayout._hasCartesian) {
+                ErrorBars.plot(gd, subplotInfo, cdError);
+                Lib.markTime('done ErrorBars');
+            }
+        }
+
+        // now draw stuff not on subplots (ie, pies)
+        // TODO: gotta be a better way to handle this
+        var cdPie = getCdModule(calcdata, Pie);
+        if(cdPie.length) Pie.plot(gd, cdPie);
+
+        // styling separate from drawing
+        Plots.style(gd);
+        Lib.markTime('done plots.style');
+
+        // show annotations and shapes
+        Shapes.drawAll(gd);
+        Annotations.drawAll(gd);
+
+        // source links
+        Plots.addLinks(gd);
+
+        return Plots.previousPromises(gd);
+    }
+
+    function cleanUp() {
+        // now we're REALLY TRULY done plotting...
+        // so mark it as done and let other procedures call a replot
+        gd._replotting = false;
+        Lib.markTime('done plot');
+        gd.emit('plotly_afterplot');
+    }
+
+    var donePlotting = Lib.syncOrAsync([
+        Plots.previousPromises,
+        marginPushers,
+        layoutStyles,
+        marginPushersAgain,
+        positionAndAutorange,
+        drawAxes,
+        drawData
+    ], gd, cleanUp);
+
+    // even if everything we did was synchronous, return a promise
+    // so that the caller doesn't care which route we took
+    return (donePlotting && donePlotting.then) ?
+        donePlotting : Promise.resolve(gd);
+};
+
+
+
+// Get the container div: we store all variables for this plot as
+// properties of this div
+// some callers send this in by DOM element, others by id (string)
+function getGraphDiv(gd) {
```

Comment:
we'll have to make `getGraphDiv` a separate module. It is used in other `plot_api.js` methods.